base/iterators.jl

```diff
@@ -739,11 +739,18 @@ julia> collect(Iterators.flatten((1:2, 8:9)))
 flatten(itr) = Flatten(itr)
 
 eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
-iteratorsize(::Type{Flatten{I}}) where {I} = SizeUnknown()
 iteratoreltype(::Type{Flatten{I}}) where {I} = _flatteneltype(I, iteratoreltype(I))
 _flatteneltype(I, ::HasEltype) = iteratoreltype(eltype(I))
 _flatteneltype(I, et) = EltypeUnknown()
 
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{T}) where {T<:Tuple} = HasLength()
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{T}) where {T<:Number} = HasLength()
+flatten_iteratorsize(a, b) = SizeUnknown()
+flatten_length(f, ::Type{T}) where {T<:Tuple} = isleaftype(T) ? nfields(T)*length(f.it) : throw(ArgumentError("Cannot compute length of a tuple-type which is not a leaf-type"))
```

Comment:
Line length here as well. Perhaps write this method definition long-form with an `if`-`else`?